src/app/clusters/microwave-oven-control-server/microwave-oven-control-server.cpp

```diff
@@ -0,0 +1,330 @@
+    }
+    mPowerSettng = powerSetting;
+}
+
+CHIP_ERROR Instance::Read(const ConcreteReadAttributePath & aPath, AttributeValueEncoder & aEncoder)
+{
+    ChipLogError(Zcl, "OperationalState: Reading");
+    switch (aPath.mAttributeId)
+    {
+    case MicrowaveOvenControl::Attributes::CookTime::Id: {
+        ReturnErrorOnFailure(aEncoder.Encode(GetCookTime()));
+    }
+    break;
+
+    case MicrowaveOvenControl::Attributes::PowerSetting::Id: {
+        ReturnErrorOnFailure(aEncoder.Encode(GetPowerSetting()));
+    }
+    break;
+
+    case MicrowaveOvenControl::Attributes::MinPower::Id: {
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetMinPower()));
+    }
+    break;
+
+    case MicrowaveOvenControl::Attributes::MaxPower::Id: {
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetMaxPower()));
+    }
+    break;
+
+    case MicrowaveOvenControl::Attributes::PowerStep::Id: {
+        ReturnErrorOnFailure(aEncoder.Encode(mDelegate->GetPowerStep()));
+    }
+    break;
+    }
+    return CHIP_NO_ERROR;
+}
+
+void Instance::InvokeCommand(HandlerContext & handlerContext)
+{
+    ChipLogDetail(Zcl, "MicrowaveOvenControl: InvokeCommand");
+    switch (handlerContext.mRequestPath.mCommandId)
+    {
+    case Commands::SetCookingParameters::Id:
+        ChipLogDetail(Zcl, "MicrowaveOvenControl:  Entering SetCookingParameters");
+
+        CommandHandlerInterface::HandleCommand<Commands::SetCookingParameters::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleSetCookingParameters(ctx, req); });
+        break;
+
+    case Commands::AddMoreTime::Id:
+        ChipLogDetail(Zcl, "MicrowaveOvenControl:  Entering AddMoreTime");
+
+        CommandHandlerInterface::HandleCommand<Commands::AddMoreTime::DecodableType>(
+            handlerContext, [this](HandlerContext & ctx, const auto & req) { HandleAddMoreTime(ctx, req); });
+        break;
+    }
+}
+
+void Instance::HandleSetCookingParameters(HandlerContext & ctx, const Commands::SetCookingParameters::DecodableType & req)
+{
+    ChipLogDetail(Zcl, "MicrowaveOvenControl:  HandleSetCookingParameters");
+    Status status;
+    uint8_t opState;
+    auto & CookMode     = req.cookMode;
+    auto & CookTime     = req.cookTime;
+    auto & PowerSetting = req.powerSetting;
+
+    // TODO(#30609): check if the input cook mode value is invalid
+
+    if (CookTime.HasValue() && (!IsCookTimeInRange(CookTime.Value())))
+    {
+        status = Status::InvalidCommand;
+        ChipLogError(Zcl, "Failed to set cookTime, cookTime value is out of range");
+        goto exit;
+    }
+
+    if (PowerSetting.HasValue() &&
+        (!IsPowerSettingInRange(PowerSetting.Value(), mDelegate->GetMinPower(), mDelegate->GetMaxPower())))
+    {
+        status = Status::InvalidCommand;
+        ChipLogError(Zcl, "Failed to set cookPower, cookPower value is out of range");
+        goto exit;
+    }
+
+    // Get Operational State instance
```

Comment:
Where is the check that at least one of CookMode, CookTime, PowerSetting is present?

Reply:
You mean we should check if there should be at least one attribute when calling the set-cooking-parameters command? 

But from the spec that the 3 fields could be missing in the command, in this case, the fields would be set to default value.

Reply:
The spec says that there are three fields, all optional, but at least one of the three must be present.